195 drag handle

6a68 opened the pull request:
@meandavejustice Here's the updated drag handle PR.

Looks like this:

<img src="https://www.dropbox.com/s/acdulovp0c38gak/Screen%20Shot%202016-09-09%20at%2010.51.08%20AM.png?dl=0&raw=true" width="400">

6a68 commented:
Here's the panel with the volume slider visible:

<img src="https://www.dropbox.com/s/3knli3gm9k0emod/Screen%20Shot%202016-09-09%20at%2010.55.33%20AM.png?dl=0&raw=true" width="400">

6a68 commented:
Note that I'm not highlighting the SVG on hover. We _could_ add a filter there, but I think it would make dragging less smooth for marginal gains.

Leaving jank investigation to a separate issue; I've tried setting a variety of timeouts, but none of them consistently help. And it seems to look pretty smooth, unless I've reloaded many times via jpm watchpost.

6a68 commented:
Hmm, looks like this change breaks the minimize button. I'll file a followup bug for that

meandavejustice commented:
@6a68 looks like this needs a rebase

meandavejustice commented on index.js:
```diff
@@ -25,12 +25,34 @@ const panel = require('sdk/panel').Panel({
   }
 });
 
-getActiveView(panel).setAttribute('noautohide', true);
```
'noautohide' needs to live outside of the `makePanelDraggable` call to ensure the panel doesn't disappear on click for linux users

6a68 replied:
good catch. I'll put that back

pdehaan commented:
I can pull this and build locally, but it looks like the font family is pretty weird in those screenshots above. Looks very "Times New Roman".

6a68 commented:
Thanks for taking a look, @pdehaan. I've noticed some busted-looking CSS as well.

The utils.js error is thrown by SDK because we change the panel, then the SDK code tries to style missing elements, assuming they exist. Probably not something we can fix.

pdehaan commented:
ANother minor observation is that after dragging the video to a new location, then resizing the browser window, the min-vid player snaps back to the lower-left corner.

6a68 commented:
> ANother minor observation is that after dragging the video to a new location, then resizing the browser window, the min-vid player snaps back to the lower-left corner.

@pdehaan On what OS are you seeing this? I'm not seeing it on my mac / dev edition.

pdehaan commented:
> On what OS are you seeing this? I'm not seeing it on my mac / dev edition.

![about_this_mac](https://cloud.githubusercontent.com/assets/557895/18453141/be69c884-78f2-11e6-8669-9f6c5f55ea7f.png)

I think I was in FF Nightly, but not sure of the exact version/date. I can re-pull your latest branch and retest if you want.

6a68 commented:
@pdehaan That would be great. Mind giving it a try in current dev edition, too?

6a68 commented:
@pdehaan Wait, never mind. I can repro the issue you mentioned now. THanks a lot for pointing it out

6a68 commented:
I found a couple other bugs that seem layout-related.

- after dragging the min-vid panel anywhere but the default location, if the window itself is moved, the panel vanishes for a few seconds (possibly while it's redrawn?). Sometimes it vanishes for a really long time, until the main window is focused and moved around a little bit, which causes XUL to actually relayout the panel.

- after minimizing but not dragging the min-vid panel, if the window is resized, the controls will be moved back upwards to the wrong spot (the controls will be displayed as though the panel was fully visible). Seems like the panel height change hasn't registered, even if it's been 5+ seconds before the window is resized.

I'm not sure if these are problems due to XUL layout limitations, or if I'm setting the height of the panel incorrectly. I'm setting the `height` attribute on both the iframe and the panel to minimize / maximize.

pdehaan commented:
My eyes also see that when you hover over the *move-icon* all the other controls disappear (which I assume is on purpose), but due to dragging lag when I drag the video player window around quickly, it's possible that my mouse speed is faster than the repaint speed, so I'm seeing controls fade in and out as I'm dragging that video popup quickly.

Sorry, hard to explain in words, but I don't know of a good way to capture a video/GIF screenshot and post it here.

![min-vid-pr214](https://cloud.githubusercontent.com/assets/557895/18453805/d0ec6658-78f5-11e6-9826-3450946b69b6.gif)

pdehaan commented:
**PROTIP:** LICEcap is great for capturing straight-to-GIF: http://www.cockos.com/licecap/

6a68 commented:
It seems like some of these bugs are only reproducible after I've had watchpost install and uninstall the addon a number of times.

pdehaan commented:
I'm seeing these on a basic `$ npm run package` then `$ ./node_modules/.bin/jpm run -b /Applications/FirefoxNightly.app`

Not sure why certain videos really look sketchy in the min-vid player, but this one seems to be breaking for me: https://www.youtube.com/watch?v=qkpyd2lr4dI

Oddly, not all the other YouTube videos from https://www.reddit.com/r/videos do that, so not sure if that's our code or youtube.

6a68 commented:
@pdehaan @meandavejustice Mind taking another look? I've updated the drag / minimize code to just track the offset relative to browser's bottom left corner. It's simpler and seems a lot more robust.

6a68 commented on index.js:
```diff
@@ -25,12 +25,39 @@ const panel = require('sdk/panel').Panel({
+
+  frame.setAttribute('height', newHeight);
+  stack.setAttribute('height', newHeight);
+  xulPanel.sizeTo(320, newHeight);
```
I think panel resizing should be very straightforward, now that I've figured out this bit

meandavejustice commented on index.js:
```diff
@@ -25,12 +25,39 @@ const panel = require('sdk/panel').Panel({
+  const frame = xulPanel.getElementsByTagName('iframe')[0];
+
+  frame.setAttribute('height', newHeight);
+  stack.setAttribute('height', newHeight);
```
This breaks on linux.

I suggest a check above where we call makePanelDraggable.

```js
let isLinux = true;
// Draggability seems to work for windows and mac, but not linux.
if (system.platform === 'winnt' || system.platform === 'darwin') {
  isLinux = false;
  makePanelDraggable(panel);
}
```

then in `adjustHeight`

```js
function adjustHeight(newHeight) {
  const xulPanel = getActiveView(panel);
  const frame = xulPanel.getElementsByTagName('iframe')[0];

  frame.setAttribute('height', newHeight);
  if (!isLinux) {
    xulPanel.getElementsByTagName('stack')[0].setAttribute('height', newHeight);
  }

  xulPanel.sizeTo(320, newHeight);

  redrawPanel();
}
```

6a68 replied:
@meandavejustice Interesting. I'll try on my linux machine. I don't really see why the stack element can't be sized. Could you paste the error?

6a68 replied:
Oh, never mind. The stack is only created if you makePanelDraggable.

I can just loop over the child elements when I adjustHeight

6a68 replied:
@meandavejustice Added some checks. Give it another try?

meandavejustice commented:
@6a68 Would you mind posting a screenshot of the player minimized?

It looks a little offset on my machine.

![minimized-min-vid-pr-214](https://cloud.githubusercontent.com/assets/1844554/18483421/efd5d2d6-79b1-11e6-910e-a042588d9872.png)

6a68 commented:
@meandavejustice Here's my player, minimized:

<img src="https://www.dropbox.com/s/c7i2cex752x4l7e/Screen%20Shot%202016-09-13%20at%2010.09.15%20AM.png?dl=0&raw=true" width="400">

pdehaan commented on index.js:
```diff
@@ -46,25 +73,13 @@ panel.port.on('addon-message', opts => {
   } else if (title === 'close') {
     panel.port.emit('set-video', {domain: '', src: ''});
+    // If the panel has been minimized, reset it to full height.
+    adjustHeight(180);
```
Should probably make `180` a constant since we're using it 3x in this page.
And on that note, we should probably do a constant for width of 320 since we use it 2x.

6a68 replied:
:+1: that'll be a defualt number that changes once resizing is added

pdehaan commented:
When dragging a "maximized" window all the other controls disappear when you hover over the drag handle. When dragging a "minimized" window, all the controls remain.

![qi___david_mitchell_s_best_moments_-_youtube](https://cloud.githubusercontent.com/assets/557895/18484143/10270fc0-799c-11e6-9174-e1a8a49dc7fe.png)

![qi___david_mitchell_s_best_moments_-_youtube](https://cloud.githubusercontent.com/assets/557895/18484175/370be034-799c-11e6-9a52-c14ba35aad03.png)

pdehaan commented:
The font family styling still seems to not be applied. Not sure if that is related to this error in my console when I start up via `jpm run`:

```sh
$ ./node_modules/.bin/jpm run -b /Applications/FirefoxNightly\ 3.app

JPM [info] Starting jpm run on Min Vid
JPM [info] Creating a new profile
console.error: min-vid:
  Unable to apply panel style
console.error: min-vid:
  Message: TypeError: Argument 1 of Window.getComputedStyle is not an object.
  Stack:
    style@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/panel/utils.js:401:55
onContentReady@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/panel/utils.js:304:7
EventListener.handleEvent*make@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/panel/utils.js:335:3
setup@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/panel.js:171:16
@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/core/disposable.js:45:56
dispatch@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/method/core.js:119:12
Disposable<.initialize@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/core/disposable.js:135:17
constructor@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/core/heritage.js:146:23
@resource://gre/modules/commonjs/toolkit/loader.js -> resource://min-vid/index.js:17:15
evaluate@resource://gre/modules/commonjs/toolkit/loader.js:279:19
load@resource://gre/modules/commonjs/toolkit/loader.js:331:5
main@resource://gre/modules/commonjs/toolkit/loader.js:786:10
run@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/addon/runner.js:147:19
startup/</<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/addon/runner.js:87:9
Handler.prototype.process@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:937:23
this.PromiseWalker.walkerLoop@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:816:7
Promise*this.PromiseWalker.scheduleWalkerLoop@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:747:11
this.PromiseWalker.schedulePromise@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:779:7
this.PromiseWalker.completePromise@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:714:7
handler@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/addon/window.js:56:3
EventListener.handleEvent*@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/addon/window.js:54:1
evaluate@resource://gre/modules/commonjs/toolkit/loader.js:279:19
load@resource://gre/modules/commonjs/toolkit/loader.js:331:5
_require@resource://gre/modules/commonjs/toolkit/loader.js:671:16
require@resource://gre/modules/commonjs/toolkit/loader.js:625:12
startup/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/addon/runner.js:72:21
Handler.prototype.process@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:937:23
this.PromiseWalker.walkerLoop@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:816:7
Promise*this.PromiseWalker.scheduleWalkerLoop@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:747:11
this.PromiseWalker.schedulePromise@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:779:7
this.PromiseWalker.completePromise@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:714:7
listener/<@resource://gre/modules/sdk/system/Startup.js:52:46
```

pdehaan commented:
Also seeing some corrupted videos (see screenshot above and below -- https://www.youtube.com/watch?v=yZT4Oi5pzg0&list=RDqkpyd2lr4dI&index=11):

![would_i_lie_to_you-_lee_mack_owes_david_mitchell_a_pen_-_youtube](https://cloud.githubusercontent.com/assets/557895/18484740/8b37cce8-799e-11e6-999c-db2281db8736.png)

I'll try and dig a bit deeper into this and see if I can repro using a locally built master versus this branch to see if I should file a separate bug.

6a68 commented:
@pdehaan Weird, that video looks fine to me:

<img src="https://www.dropbox.com/s/zpp8xoc7uhl9bxc/Screen%20Shot%202016-09-13%20at%2010.42.59%20AM.png?dl=0&raw=true" width="400">

pdehaan commented on index.js:
```diff
@@ -4,33 +4,71 @@
+  // On linux, makePanelDraggable isn't invoked, so the stack doesn't exist.
+  const xulPanel = getActiveView(panel);
+  const stack = xulPanel.getElementsByTagName('stack') && xulPanel.getElementsByTagName('stack')[0];
+  const frame = xulPanel.getElementsByTagName('iframe')[0];
```
Will there always be an `<iframe>`?
I see on L51, we do a check for `stack` before trying to access `stack[0]`, but here we just directly access `iframe[0]`

6a68 replied:
Yeah, there's always an iframe. The stack is a weird XUL-only thing required to absolutely position the drag handle over top of the iframe.

6a68 commented:
Think I've fixed the CSS errors and the `Argument 1 of Window.getComputedStyle is not an object` as well:

<img src="https://www.dropbox.com/s/16cdyswxmc424t6/Screen%20Shot%202016-09-13%20at%2011.20.55%20AM.png?dl=0&raw=true" width="400">

Added a setTimeout to give the SDK code a chance to finish styling the panel before we change its `type` (which removes some anonymous elements that the SDK expects to exist).

pdehaan commented:
@6a68 Looking good!

I still see some elements jump in and out when dragging (if I move fast enough), and a weird glitch when the video is finished and you minimize the player, but I imagine both of those could be filed and triaged as new issues.

![_r_videos_and_195_drag_handle_by_6a68_ _pull_request__214_ _meandavejustice_min-vid](https://cloud.githubusercontent.com/assets/557895/18486611/9cb44f16-79a6-11e6-95c2-36fe300eb871.png)

versus

![_r_videos_and_195_drag_handle_by_6a68_ _pull_request__214_ _meandavejustice_min-vid](https://cloud.githubusercontent.com/assets/557895/18486625/abfa1014-79a6-11e6-9ddd-1c1b7764abcb.png)

If @meandavejustice is happy, I'm happy.

6a68 commented:
@pdehaan Fantastic. Thanks for your help with testing! :beers:

pdehaan commented:
@6a68 Let me know if you want me to file the 2 outstanding issues ("dragging fast causes UI icons to flicker" and "minimized player shows replay/close icons when video stops playing") in https://github.com/meandavejustice/min-vid/pull/214#issuecomment-246780452.

meandavejustice commented:
@pdehaan You can file those issues, that would be v helpful

meandavejustice commented on index.js:
```diff
@@ -4,33 +4,71 @@
+  const frame = xulPanel.getElementsByTagName('iframe')[0];
+
+  frame.setAttribute('height', newHeight);
+  if (stack) { stack.setAttribute('height', newHeight); }
```
nit: can you get rid of these braces?

6a68 replied:
:+1:
